src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.cs

```diff
@@ -241,7 +240,7 @@ private void EnsureIOInitialized()
             // If at this point request was not upgraded just start a normal IO engine
             if (AsyncIO == null)
             {
-                AsyncIO = new AsyncIOEngine(_pInProcessHandler);
+                AsyncIO = new AsyncIOEngine(_contextLock, _pInProcessHandler);
```

Comment:
Seems bad to have a shared lock between the HttpContext and AsyncIOEngine. Worth it to move OnStarting and OnCompleted to the AsyncIOEngine?

Reply:
`OnStarting` and `OnCompleted` don't really belong to IOEngine having multiple lock objects might have been more correct but it's just more allocations.